### ISSUE TITLE
Feature request: File Handling Improvements

### ISSUE THREAD

**Issue by berntpopp** [hassansaei/VNtyper#20]
## Problem

`vntyper pipeline` has no way to reuse a completed step. Every run recomputes everything, and nothing warns when the output directory is non-empty.

The checkpoint file the 2024 work plan proposed already exists. `pipeline_summary.json` is re-serialised after every step with `step`, `start`, `end`, `command`, `result_file`, `file_type`, `md5sum`, `parsed_result` and `result_file_missing`, and `pipeline_end` is written only on the success path. Nothing reads it.

Separately, `--keep-intermediates` is accepted by `cli_parser.py`, documented, and read nowhere. An advertised no-op is its own defect.

Research use only.

## Deliverable

1. **Remove or wire `--keep-intermediates`.** Small, first, on its own.
2. **A reader, not a second manifest.** `vntyper/scripts/resume.py` exposing `step_is_complete(summary, step_name) -> bool`, using the canonical names in `summary_steps.py`; no new string literals (AGENTS.md trap 5). Prerequisite: the remaining inline step-name literals in `pipeline.py` become constants.
3. **`--resume`** skips a step only when the manifest marks it complete **and** `result_file` still matches its recorded `md5sum`. Existence is not sufficient (#212, #223).
4. **`--force`** ignores the manifest. Default stays today's behaviour, plus a warning on a non-empty output directory.
5. **Refuse `--resume`** when the recorded `version` differs from `vntyper.version.__version__`, when `input_files` differs, or when the recorded decision profile differs (#306 records it). Resuming across a recalibration is the silent-wrong-answer this pipeline must not produce.
6. Scope the flag to `vntyper pipeline`. `vntyper online` already has a `--resume` that resumes remote polling.

## Sequencing

After **#314**. A resume that trusts a file at the public path is only safe once an interrupted subprocess cannot leave a partial file there.

## Out of scope

Prompting, corruption repair, partial-file recovery, a separate status file, and `install-references` reuse (digest-gated since milestone 5).

## History

Original body, 2024-11-17: "Clean up logic to allow for skipping, resuming, or overwriting existing files, providing better control over pipeline execution." The 2024-12-10 work plan and the 2026-08-07 and 2026-08-12 audits are in the comments. This body supersedes them.

**Comment by berntpopp** [comment]
**Work Plan for File Handling Improvements**

**Additional Consideration: Pipeline Status and Checkpointing**  
Beyond simply checking if files exist, it’s crucial to verify the integrity or successful completion of the steps that generated those files. A file’s mere presence does not guarantee correctness (the file could be corrupted, truncated, or from a previous partial run). To handle this, we can introduce a **status checkpoint system**:

1. **Checkpoint Files or State Records:**  
   After each step in the pipeline completes successfully, create or update a small JSON or YAML file (e.g., `pipeline_status.json`) in the output directory. This file would record:
   - Which steps have completed successfully.
   - Relevant timestamps.
   - Checksums or sizes of key output files (if feasible) to ensure data integrity.
   
   For example:
   ```json
   {
     "steps": {
       "fastq_qc": {
         "status": "completed",
         "timestamp": "2024-12-10T12:34:56Z",
         "files": {
           "output_R1.fastq.gz": "md5_or_size_info",
           "output_R2.fastq.gz": "md5_or_size_info"
         }
       },
       "bam_to_fastq": {
         "status": "completed",
         "timestamp": "2024-12-10T12:50:11Z"
       },
       "kestrel_genotyping": {
         "status": "completed",
         "timestamp": "2024-12-10T13:10:03Z"
       }
     }
   }
   ```

   The pipeline can consult this file at the beginning of each run (or step) to determine if previous steps were completed successfully and the files remain consistent. This prevents re-running completed steps unless the user requests overwriting or redoing them.

2. **Corruption or Inconsistency Detection:**  
   If we store file hashes or sizes, on `--resume` or `--skip-existing` mode, the pipeline can re-check these files:
   - If the file hash or size doesn’t match what was recorded, the pipeline will know the file may be corrupted or changed, prompting the user to overwrite or re-run the step.
   - If no recorded status is found, the pipeline cannot trust that step was completed, so it must run that step unless instructed otherwise.

3. **Fallback to Logs:**  
   While reading logs to deduce status is possible, logs aren’t structured for reliable machine parsing. The dedicated checkpoint file is cleaner. We can still append a note in logs when we create/update the checkpoint, but the primary machine-readable status would come from the checkpoint file.

4. **User-Controlled Behavior:**  
   With `--skip-existing`, the pipeline checks the checkpoint file and existing outputs:
   - If step outputs are present and verified, skip that step.
   - If `--resume` is used, attempt to resume from the last completed step according to the checkpoint.
   - If `--overwrite` is used, regenerate even if the checkpoint says it’s done.

**Proposed Work Plan**

1. **Add Command-Line Flags:**
   - `--skip-existing`: Skip steps if their outputs (and checkpoints) are valid.
   - `--resume`: Resume pipeline from where it last left off based on the checkpoint file.
   - `--overwrite`: Force regeneration of steps/files regardless of existing outputs.

2. **Implement a Checkpoint System:**
   - After each major pipeline step completes (e.g., `process_fastq`, `process_bam_to_fastq`, `run_kestrel`), create or update a JSON file (`pipeline_status.json`) in the output directory.
   - Record step name, timestamp, file details (hash/size), and “completed” status.
   - On the next run, read this file at startup to determine which steps are already completed.

3. **File Integrity Checks:**
   - Optionally compute and store MD5 checksums or file sizes of critical output files when completing a step.
   - On the next run with `--resume` or `--skip-existing`, re-check these files to ensure they match. If mismatched, consider them invalid and re-run the step (or prompt the user).

4. **Integrate into Existing Steps:**
   - Modify `run_pipeline` to use the checkpoint file. Before running each step:
     - Check if the step is marked “completed” and if files match expected checksums.
     - If yes and `--skip-existing` or `--resume` is set, skip running the step.
     - If `--overwrite` is set, run the step anyway.
   
   - Update `process_bam_to_fastq`, `process_fastq`, and `run_kestrel` (and other key steps) to:
     - Check if they can skip execution due to a completed checkpoint.
     - On successful completion, update the checkpoint file.

5. **Error Handling and Corruption:**
   - If a step’s files differ from what the checkpoint recorded, log a warning. If the user has `--resume` or `--skip-existing`, prompt them to use `--overwrite` or re-run without skip/resume.
   - Document this behavior in the CLI help and user guide.

6. **Testing:**
   - Test scenarios:  
     - Running pipeline fully, then re-running with `--resume`.
     - Corrupting a file and checking that the pipeline refuses to skip.
     - Using `--overwrite` to forcibly re-run steps even if already completed.
     - Ensuring `--skip-existing` and `--keep-intermediates` interact nicely.

7. **Documentation:**
   - Update the README or user manual to explain the new flags and the checkpoint file.
   - Advise users that if they modify output files outside the pipeline or experience partial runs, the checkpoint may indicate stale information. In such cases, use `--overwrite`.

---

By implementing this checkpointing approach and integrating it with the new flags, we will achieve a more robust, user-friendly, and reliable file handling system that allows skipping, resuming, or overwriting steps as desired, while maintaining confidence in the integrity of the pipeline’s outputs.

**Comment by berntpopp** [comment]
Audited at `main` @ `39b4a54`. Keeping this open, but two things change its shape.

## 1. The unsafe version of this feature already exists, unflagged

`vntyper/scripts/kestrel_genotyping.py:294-297` skips the entire Kestrel stage whenever
`output.vcf` exists — unconditional, gated on no flag, untested, undocumented, and because it
`return`s it also skips `process_kestrel_output`, so `kestrel_result.tsv` is never
regenerated. Re-running into a directory left by an interrupted run can therefore produce a
**silent negative**.

Filed as #212. It should be fixed before, or as part of, any resumability work — not left in
place underneath it.

## 2. The checkpoint file in the work plan above already exists

The plan proposes `pipeline_status.json` recording per-step status, timestamps and checksums.
`pipeline_summary.json` already is that file. `record_step` is called with
`write_summary_path=summary_file_path` at all ten step sites (`pipeline.py:302, 332, 369,
390, 415, 448, 484, 519, 601, 628`) and `summary.py:238-239` re-serialises after **every**
step, carrying per step: name, start, end, command, `result_file`, `file_type` and `md5sum`.
A partial run is already distinguishable from a complete one — `pipeline_end` is written only
on the success path (`pipeline.py:698-699`).

Roughly 70 % of the proposed system is on disk. What is missing is a **reader**. Do not add a
second manifest.

## What exists today for resumability

Effectively nothing. No `--resume`/`--force`/`--overwrite`/`--skip-existing` on the
`pipeline` subcommand (`cli_parser.py:107-121` has only `--keep-intermediates`,
`--delete-intermediates`, `--archive-results`). The single `--resume` in the codebase belongs
to `vntyper online` and resumes polling a remote job. Two step-level reuse checks exist, both
in `fastq_bam_processing.py` (`:135-136`, `:233-245`), both correctly gated behind
`--keep-intermediates`.

## Narrowed deliverable

- `--resume` — skip a step the manifest marks complete **and** whose `result_file` still
  matches its recorded `md5sum`.
- `--force` — ignore the manifest, re-run everything.
- Default — today's behaviour, plus a warning when the output directory is non-empty.
- One new module `vntyper/scripts/resume.py` exposing
  `step_is_complete(summary, step_name) -> bool`. `summary_steps.py` already provides
  `get_step`/`get_step_result` (`:42-72`) and the canonical names, so no new string literals
  (AGENTS.md trap 5).
- **Refuse `--resume` when the recorded `version` differs from `vntyper.version.__version__`,
  or when `input_files` differs.** Resuming across a version bump with recalibrated
  thresholds is exactly the silent-wrong-answer this pipeline must not produce.
- Fold the two existing `--keep-intermediates` reuse checks into the same mechanism, so there
  is one concept rather than two.

Out of scope: prompting the user, corruption repair, partial-file recovery, and a separate
status file.

## Honest value assessment

The **safety** half is worth more than the speed half. A default Kestrel-only run is minutes;
adVNTR adds ~9 min/sample. Resume saves real time only on adVNTR runs and cohort-scale
re-runs. But the current unconditional VCF skip can turn a dirty directory into a false
negative in a clinical pipeline — that is worth fixing on its own merits, today, which is why
#212 is separate and milestoned ahead of this.

**Comment by berntpopp** [comment]
**Superseding my 2026-08-07 audit comment on this issue: its motivating finding was fixed five
days after it was written.** A planner reading it today would chase a bug that no longer exists.
Re-verified at `main` @ `87db9a9`.

## What is now stale

The comment's headline finding — that `kestrel_genotyping.py` skips the Kestrel stage whenever
`output.vcf` exists — was removed by `0a0af2a` ("fix(kestrel): never skip the stage because
output.vcf exists", 2026-08-07, `Closes #212`). The code now does the opposite: it deletes a
stale VCF and always re-runs (`kestrel_genotyping.py:213-215`), with a `completed` flag guard at
`:239`/`:261`/`:268`. The comment at `:207` explicitly defers deliberate reuse to "the
`--resume` flag proposed in #20".

The comment's second survey is stale too: the two `--keep-intermediates`-gated reuse checks in
`fastq_bam_processing.py` were removed by `f50db4e`.

**Net: there is now zero step-level reuse anywhere in the run path.** No `--resume`, `--force`
or `--overwrite` on `pipeline`, and no warning when the output directory is non-empty.

## What still holds

Both structural claims survive, and they are the reason this issue should not grow a second
manifest:

- **`pipeline_summary.json` already is the proposed checkpoint file.** Written at
  `pipeline.py:269`; `summary.py:302-315` re-serialises the whole dict after **every** step.
  Per step (`summary.py:261-270`): `step`, `start`, `end`, `command`, `result_file`,
  `file_type`, `md5sum` (set at `:285`), `parsed_result`, and `result_file_missing`.
  **Exactly 10 call sites, all passing `write_summary_path`** — `pipeline.py:305, 316, 350, 372,
  397, 446, 472, 559, 586` and `pipeline_kestrel.py:76`. Zero omissions, corroborated by
  `tests/unit/test_summary_step_contract.py:328`. (One correction to my own comment: the key is
  `step`, not `name`.)
- **A partial run is already distinguishable.** `pipeline_end` is set only in `end_summary()`
  (`summary.py:99`), called once at `pipeline.py:631` inside the `try` — absent from both
  `except` arms and the `finally`. No reader gates on it yet.

## What a rewritten issue should ask for

1. **A reader, not a second manifest.** The checkpoint file exists; nothing consumes it.
2. **`--resume` skips a step only when the manifest marks it complete *and* `result_file` still
   matches its recorded `md5sum`.** Existence is not sufficient — that is exactly the #212
   failure mode, and #223 shows the same shape again with a well-formed-looking file.
3. **`--force`, plus a warning on a non-empty output directory.**
4. **Refuse `--resume` across a `version` or `input_files` change.** Resuming over a threshold
   recalibration is precisely the silent-wrong-answer this pipeline must not produce. Both
   fields are already in the summary.
5. **Resolve `--keep-intermediates`, which is now an advertised no-op** — accepted at
   `fastq_bam_processing.py:91`, documented at `:109`, threaded from `cli_parser.py:108` →
   `cli_handlers.py:453` → `pipeline.py:291, 439`, and **never read**. Give it meaning or remove
   it; an inert documented flag is its own defect.

**Prerequisite:** only 5 of the 10 step names are constants (`summary_steps.py:25-29`). Promote
the inline literals at `pipeline.py:316, 350, 372, 397, 446` first, or a resume reader
reintroduces the string-literal problem AGENTS.md records as trap 5.

**Naming collision to avoid:** `online_mode.py:294` already has a `--resume`, and it resumes
*remote job polling*. Pick a different name or scope this one explicitly to `vntyper pipeline`.

## Sequencing

Do not schedule ahead of **#223**. A `--resume` that trusts recorded step completion is only
safe once completion is recorded honestly, and #223 is an open case where a step completes,
records a result file, and the result is a manufactured negative.

## Adjacent, worth its own issue

`install_references.py:78-80` skips a download whenever the destination exists — no checksum, no
size check, no override. Same silent-corruption shape as #212, one directory over. And
`force_reindex` (`:342`) has no CLI exposure.

---

Moved out of the milestone structure to backlog until the body is rewritten along these lines.

**Comment by berntpopp** [comment]
Owner decision 2026-09-02: `--keep-intermediates` is fixed now as its own small change (parsed at `cli_parser.py:203`, read nowhere). The resume mode is sequenced after #314, because a manifest reader may only trust a file at the public path once an interrupted subprocess cannot leave a partial one there. Body rewritten to the narrowed deliverable from the 2026-08-12 audit.
